File: dart/cloudtest/lib/cloudtest.dart

import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:path/path.dart';
import 'package:http/http.dart' as http;
import 'package:async/async.dart';

String serviceEndpoint = '';
String apiKey;
HttpClient client = HttpClient();

Future readStackOutputs() async {
  var stackOutputs = await(File('../../stackoutputs.json')
    .readAsString()
    .then((fileContents) => json.decode(fileContents)));
  apiKey = stackOutputs['testapikey'];
  serviceEndpoint = stackOutputs['ServiceEndpoint'];
}

Future<Map> apiClientGet(String relativePath) async {
  assert(serviceEndpoint.isNotEmpty);
  var url = serviceEndpoint + relativePath;
  var response = await http.get(url, headers: {'X-Api-Key': apiKey});
  if (response.statusCode != 200) {
    throw Exception('Get ${relativePath} failed: ${response.statusCode}');
  }
  return json.decode(response.body);
}

Future<Map> apiClientPost(String relativePath, Map data) async {
  assert(serviceEndpoint.isNotEmpty);
  var url = serviceEndpoint + relativePath;
  var response = await http.post(
    url,
    body: jsonEncode(data),
    headers: {'X-Api-Key': apiKey, 'Content-Type': 'application/json'});
  if (response.statusCode != 200) {
    throw Exception('Post ${relativePath} failed: ${response.statusCode}');
  }
  return json.decode(response.body);
}


Future<String> getManifestUrl() async {
  var response = await apiClientGet('/api/v1/manifestUrl');
  return response['url'];
}

Future<Map> getManifest() async {
  var url = await getManifestUrl();
  var response = await http.get(url);
  if (response.statusCode != 200) {
    throw Exception('Get manifest failed: ${response.statusCode}, ${url}');
  }
  return json.decode(response.body);
}

Future destroyAllData() async {
  await apiClientPost('/fortesting/internal/destroyalldata', {});
}

Future uploadMediaFile(String path) async {
  var pathBasename = basename(path);
  var presignedPost = await apiClientPost('/api/v1/requestFileUpload', {'filename': pathBasename});
  var url = presignedPost['presignedPost']['url'];
  Map<String, dynamic> fields = presignedPost['presignedPost']['fields'];

  var imageFile = File('testdata/test.png');

  var stream = http.ByteStream(imageFile.openRead());
  var length = await imageFile.length();

  var uri = Uri.parse(url);

  var request = http.MultipartRequest('POST', uri);
  var multipartFile = http.MultipartFile('file', stream, length, filename: basename(imageFile.path));
          //contentType: new MediaType('image', 'png'));

  fields.forEach((key, value) {request.fields[key] = value;});
  request.files.add(multipartFile);
  var response = await request.send();
  if (response.statusCode < 200 && response.statusCode >= 300) {
    var body = await response.stream.bytesToString();
    throw Exception('Post to upload file failed: ${response.statusCode}: ${body}');
  }
}

Future test1() async {
  await destroyAllData();
  var manifest = await getManifest();
  assert(manifest['mediaFiles'].length == 0);
  await uploadMediaFile('testdata/test.png');
  sleep(Duration(seconds: 2));
  var manifest2 = await getManifest();
  assert(manifest2['mediaFiles'].length == 1);
  String key = manifest2['mediaFiles'][0]['key'];
  assert(key.endsWith('test.png'));
  print('DONE!');
}

File: ts/tests/cloudtest.ts
import * as apiclient from '../apiclient/apiclient';
import { apiClient } from '../apiclient/apiclient';
import * as fs from 'fs';
import * as path from 'path';
import * as expect from 'expect';
import fetch from 'node-fetch';
import { postFile, fetchWrapper } from '../apiclient/nodebrowserduality';
import * as sleep from 'sleep-promise';
import { FileKind } from '../model/model';
import * as s3 from '../backend/s3';

const stackOutputs = JSON.parse(fs.readFileSync(path.join(process.env.TOP!, 'stackoutputs.json')).toString());

async function destroyAllData() {
    const response = await fetch(`${apiclient.settings.serviceEndpoint}/fortesting/internal/destroyalldata`, {
        method: 'POST',
        headers: {'X-API-Key': apiclient.settings.apiKey},
    });
    if (!response.ok) {
        throw Error(`Not OK`);
    }
    while (true) {
        const mediaFiles = await s3.list(stackOutputs.VersionsBucket);
        if (mediaFiles.length === 0) {
            break;
        }
        await s3.erase(mediaFiles, stackOutputs.VersionsBucket);
        console.warn(`Destroyed ${mediaFiles.length} media files!`);
    }
}

async function main() {
    apiclient.settings.apiKey = stackOutputs['testapikey-dev'];
    apiclient.settings.serviceEndpoint = stackOutputs.ServiceEndpoint;
    apiclient.settings.fetchWrapper = fetchWrapper;

    await destroyAllData();
    let manifestUrl = await apiClient.manifestUrl();
    let manifest = await apiClient.getManifest(manifestUrl);
    expect(manifest['mediaFiles'].length).toBe(0);
    const responseFileUpload = await apiClient.requestFileUpload({filename: 'test.png', kind: FileKind.GALLERY_MEDIA});
    const {url, fields} = responseFileUpload['presignedPost'];
    await postFile(url, fields, path.join(process.env.TOP!, 'testdata', 'test.png'));
    await sleep(2000);
    manifestUrl = await apiClient.manifestUrl();
    manifest = await apiClient.getManifest(manifestUrl);
    expect(manifest.mediaFiles.length).toBe(1);
    expect(manifest.mediaFiles[0].key.endsWith('test.png')).toBeTruthy();
    const imageUrl = manifest.mediaFiles[0].url;
    const response = await fetch(imageUrl);
    expect(response.ok).toBeTruthy();
    console.log('DONE!');
}

main();
